### ISSUE TITLE
Thread can never execute the query in the database

### ISSUE THREAD

**Issue by Moliholy** [SQLDroid/SQLDroid#52]
Hello, and first of all thanks for this awesome library.

I have an issue when executing PreparedStatements inside an AsyncTask. Firstly I execute a normal statement, and secondly I try to execute a PreparedStatement. The first statement is successfully created, but the second one stays forever in the following line:

```
                db.getThreadSession().prepare(mSql,
                        db.getThreadDefaultConnectionFlags(assumeReadOnly),
                        cancellationSignalForPrepare, info);
```java
android.database.sqlite.SQLiteProgram.java, line 58, concretely.

Any clue about this?
Thanks a lot in advance

**Comment by donv** [comment]
Hi @Moliholy !

This sounds like a multithreading locking issue.  If you could try to dump the threads when the app is locked, that would give a bit more information.

**Comment by Moliholy** [comment]
As far as I've seen it looks like the AsyncTask threads are not destroyed immediately after finishing their job. That means that if the database is opened in RW mode (which actually is the default behaviour of the library) then other threads won't be able to access it while the first thread lives due to the android sharing policy in RW database connections, which states that multithreading in RO connections is possible, but not in RW connections, where only one thread can hold it.

**Comment by donv** [comment]
Sounds like something we should handle.  Do you have time to look at the code and do a pull request?  I'd be happy to merge and release the changes.

**Comment by Moliholy** [comment]
Of course, I'll take a look to see if I can make it work and submit a PR with the changes

**Comment by jhannes** [comment]
@Moliholy Could this be the same as my issue here: https://github.com/SQLDroid/SQLDroid/issues/54

**Comment by Moliholy** [comment]
@jhannes I'd bet it is related. Connections are always opened with write permissions, and no more than one thread can access simultaneously in these conditions. In that context, it doesn't really matter if you use autocommit or not, simply accessing in multithreaded mode is enough to reproduce the issue.
... or at least that's my guess, I might be wrong.

My final workaround was to have a unique thread that connects to many databases, which waits until a job enters in a queue. In case it's of any use to you here you have a handmade thread synchronization (and at the same time a nightmare I'd like to forget) is here: https://github.com/Moliholy/cvmfs-android/blob/devel/app/src/main/java/ch/cern/cvmfs/model/RepositoryManager.java

**Comment by jhannes** [comment]
@Moliholy I managed to make the problem dissappear for me if I did setAutoCommit(false) after commit.

If the problem returns I think your RepositoryManager may be my solution as well. Thanks for the tip.

**Comment by dperiwal** [comment]
@jhannes Did you mean to say "I managed to make the problem disappear for me if I did setAutoCommit(true) after commit"? That's what I read in your commented code in #54

**Comment by jhannes** [comment]
@dperiwal. Whops. Yes, you are correct. Auto commit is set to *true*.
